rust_src/src/xdisp.rs

```diff
@@ -1,13 +1,30 @@
 //! Display generation from window structure and buffer text.
 
-// Remove the feature check when this file gains functions
-// that are not guarded by it.
-#[cfg(feature = "glyph-debug")]
 use remacs_macros::lisp_fn;
 
+use crate::lisp::LispObject;
+use crate::remacs_sys::{bset_update_mode_line, Qnil};
+use crate::threads::ThreadState;
 #[cfg(feature = "glyph-debug")]
 use crate::{interactive::InteractiveNumericPrefix, remacs_sys::trace_redisplay_p};
 
+/// Mark the current buffer for redisplay.
+/// This function may be passed to `add-variable-watcher'.
+#[lisp_fn]
+pub fn set_buffer_redisplay(
+    _symbol: LispObject,
+    _newval: LispObject,
+    _op: LispObject,
+    _where: LispObject,
+) -> LispObject {
+    let mut current_buffer = ThreadState::current_buffer_unchecked();
```

Comment:
You are correct this looks pretty close to `force_mode_line_update`. We have choices at this point.
We can ignore it and allow some minor code duplication or we can pull this out into a function or method on LispBufferRef. Then this function becomes something like:

```
let mut current_buffer = ThreadState::current_buffer_unchecked();
current_buffer.set_buffer_redisplay();
```

and the else in `force_mode_line_update` does the same.

Reply:
Thinking about it a bit more, I'm wary of pulling the common code out. The two functions are claiming to do different things - redisplay the buffer, and redisplay the mode line. The fact that they're using the same code in the two cases suggests they're misnamed, or at least do more work than their names suggest.

It feels to me like the correct (eventual) refactor would be to make them actually do different things that match their names. Pulling out the common code into a new function works against that. 

So, I propose to leave in the duplication. What do you think?

Reply:
I don't have a problem with that since we don't really have much code duplication with this.

But I just realized that we commented out `current_buffer.set_prevent_redisplay_optimizations_p(true)` in `force_mode_line_update´.

@shaleh what was your intention for doing this ?

Reply:
A comment in the C code questioning its value. This code is a bit too complex and frankly no one, here or on Emacs side I suspect could really tell us what is happening. I commented it out as an experiment. It is fine to put it back.